### PR TITLE
Refactoring of trees

### PR DESCRIPTION
**Pull request by interesaaat**
This PR cleans up a bit the tree converters \ implementation. It also fixes few problems with the decision tree \ random forest converters for ONNX.

### CODE REVIEW

File: hummingbird/ml/operator_converters/_gbdt_commons.py [hunk 1]
```diff
@@ -75,26 +74,85 @@ def convert_gbdt_common(tree_infos, get_tree_parameters, n_features, classes=Non
...
+                extra_config[constants.POST_TRANSFORM] = lambda x: apply_sigmoid(apply_base_prediction(base_prediction)(x))
+            else:
+                extra_config[constants.POST_TRANSFORM] = apply_sigmoid
+        elif extra_config[constants.POST_TRANSFORM] == constants.SOFTMAX:
+            if constants.BASE_PREDICTION in extra_config:
+                extra_config[constants.POST_TRANSFORM] = lambda x: apply_softmax(apply_base_prediction(base_prediction)(x))
+            else:
+                extra_config[constants.POST_TRANSFORM] = apply_softmax
+        elif extra_config[constants.POST_TRANSFORM] == "NONE":
+            extra_config.pop(constants.POST_TRANSFORM)
```
Comment (ksaur) [inline hummingbird/ml/operator_converters/_gbdt_commons.py]: Can we add a test-case for NONE?

File: hummingbird/ml/operator_converters/onnx/onnxml_tree_ensemble.py [hunk 2]
```diff
@@ -90,12 +98,35 @@ def _get_tree_infos_from_onnx_ml_operator(model):
             t_right = right[prev_id:count]
             t_features = features[prev_id:count]
             t_threshold = threshold[prev_id:count]
-            t_values = [0] * len(t_left)
-            for j in range(len(t_left)):
-                if t_threshold[j] == -1 and l_count < len(values):
-                    t_values[j] = values[l_count]
+            t_values = np.zeros((len(t_left), n_classes)) if is_decision_tree else np.zeros(len(t_left))
+            if len(t_left) == 1:
+                # Model creating trees with just a single leaf node. We transform it
```
Comment (ksaur) [inline hummingbird/ml/operator_converters/onnx/onnxml_tree_ensemble.py:103]: add a test case for this?

File: hummingbird/ml/operator_converters/onnx/onnxml_tree_ensemble.py [hunk 3]
```diff
@@ -90,12 +98,35 @@ def _get_tree_infos_from_onnx_ml_operator(model):
...
+                    t_values[0] = values[l_count]
                     l_count += 1
-            tree_infos.append(TreeParameters(t_left, t_right, t_features, t_threshold, np.array(t_values).reshape(-1, 1)))
+            else:
+                for j in range(len(t_left)):
+                    if t_threshold[j] == -1 and l_count < len(values):
+                        t_values[j] = values[l_count]
+                        l_count += 1
+            if t_values.shape[0] == 1:
+                # Model creating trees with just a single leaf node. We fix the values here.
```
Comment (ksaur) [inline hummingbird/ml/operator_converters/onnx/onnxml_tree_ensemble.py:118]: test case?

File: hummingbird/ml/operator_converters/_gbdt_commons.py [hunk 4]
```diff
@@ -75,26 +74,85 @@ def convert_gbdt_common(tree_infos, get_tree_parameters, n_features, classes=Non
         for parameter in tree_parameters:
             parameter.values = parameter.values * extra_config[constants.LEARNING_RATE]
 
-    # Generate the tree implementation based on the selected strategy.
+    # Generate the network parameters based on the selected strategy.
```
Comment (ksaur) [inline hummingbird/ml/operator_converters/_gbdt_commons.py]: What do you mean by "network parameters"?

File: hummingbird/ml/operator_converters/_gbdt_commons.py [hunk 5]
```diff
@@ -75,26 +74,85 @@ def convert_gbdt_common(tree_infos, get_tree_parameters, n_features, classes=Non
...
+                tree_param.thresholds,
+                tree_param.values,
+                n_features,
+                extra_config,
             )
             for tree_param in tree_parameters
         ]
-        return GEMMGBDTImpl(net_parameters, n_features, classes, extra_config)
+    else:
+        # Some models require some additional massagging of the parameters before generating the tree_trav implementation.
```
Comment (ksaur) [inline hummingbird/ml/operator_converters/_gbdt_commons.py]: s/massagging/massaging/g

File: hummingbird/ml/operator_converters/_tree_implementations.py [hunk 6]
```diff
@@ -48,19 +47,6 @@ def aggregation(self, x):
         """
         pass
 
-    @abstractmethod
-    def calibration(self, x):
```
Comment (ksaur) [inline hummingbird/ml/operator_converters/_tree_implementations.py:52]: can you explain why this isn't needed (so that i understand)?  Is this related to #218 ?

File: tests/test_onnxml_decision_tree_converter.py [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+
+class TestONNXDecisionTreeConverter(unittest.TestCase):
+    def __init__(self, *args, **kwargs):
+        super(TestONNXDecisionTreeConverter, self).__init__(*args, **kwargs)
+
+    # Base test implementation comparing ONNXML and ONNX models.
+    def _test_decision_tree(self, X, model, extra_config={}):
+        # Create ONNX-ML model
+        onnx_ml_model = convert_sklearn(
+            model, initial_types=[("input", FloatTensorType([X.shape[0], X.shape[1]]))], target_opset=9
```
Comment (ksaur) [inline tests/test_onnxml_decision_tree_converter.py]: opset 11?  or is this one stuck on 9?

File: hummingbird/ml/operator_converters/_gbdt_commons.py (hunk 1 above)
Comment (interesaaat) [reply hummingbird/ml/operator_converters/_gbdt_commons.py]: I removed `NONE`. With `NONE` currently we actually use the `DecisionTree` implementation, not the `GBDT` one.

File: hummingbird/ml/operator_converters/onnx/onnxml_tree_ensemble.py (hunk 2 above)
Comment (interesaaat) [reply hummingbird/ml/operator_converters/onnx/onnxml_tree_ensemble.py:103]: Added.

File: hummingbird/ml/operator_converters/_gbdt_commons.py (hunk 4 above)
Comment (interesaaat) [reply hummingbird/ml/operator_converters/_gbdt_commons.py]: Changed into "model parameters"

File: hummingbird/ml/operator_converters/_gbdt_commons.py (hunk 5 above)
Comment (interesaaat) [reply hummingbird/ml/operator_converters/_gbdt_commons.py]: Good catch!

File: hummingbird/ml/operator_converters/_tree_implementations.py (hunk 6 above)
Comment (interesaaat) [reply hummingbird/ml/operator_converters/_tree_implementations.py:52]: I pushed everything into the aggregation function. In this way we don't have to use two different functions to change the behavior of trees. In calibration I was basically putting the onnx `post_transform`. DecisionTree implementations don't need post transform, while the post_transforms for GBDT implementations now are inside aggregation.

File: tests/test_onnxml_decision_tree_converter.py (hunk 7 above)
Comment (interesaaat) [reply tests/test_onnxml_decision_tree_converter.py]: good catch!

File: hummingbird/ml/operator_converters/onnx/onnxml_tree_ensemble.py (hunk 3 above)
Comment (interesaaat) [reply hummingbird/ml/operator_converters/onnx/onnxml_tree_ensemble.py:118]: Added.
